File: backend-spring-boot/src/test/java/no/vicx/backend/user/service/UserServiceTest.java
package no.vicx.backend.user.service;

import no.vicx.backend.error.NotFoundException;
import no.vicx.backend.user.vm.ChangePasswordVm;
import no.vicx.backend.user.vm.UserPatchVm;
import no.vicx.database.user.UserRepository;
import no.vicx.database.user.VicxUser;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;
import org.junit.jupiter.params.provider.ValueSource;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.cache.Cache;
import org.springframework.cache.CacheManager;
import org.springframework.mock.web.MockMultipartFile;
import org.springframework.security.crypto.password.PasswordEncoder;

import java.io.IOException;
import java.util.Optional;

import static no.vicx.backend.user.UserTestUtils.VALID_USER_VM;
import static no.vicx.backend.user.UserTestUtils.createValidVicxUser;
import static no.vicx.database.user.VicxUser.VALID_BCRYPT_PASSWORD;
import static no.vicx.database.user.VicxUser.VALID_PLAINTEXT_PASSWORD;
import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class UserServiceTest {

    @Mock
    UserRepository userRepository;

    @Mock
    PasswordEncoder passwordEncoder;

    @Mock
    CacheManager cacheManager;

    @Mock
    Cache reCaptchaCache;

    UserService sut;

    @BeforeEach
    void setUp() {
        when(cacheManager.getCache("RECAPTCHA_TOKENS")).thenReturn(reCaptchaCache);

        sut = new UserService(userRepository, passwordEncoder, cacheManager);
    }

    @Test
    void constructor_givenNoRecaptchaTokensCache_expectIllegalStateException() {
        when(cacheManager.getCache("RECAPTCHA_TOKENS")).thenReturn(null);

        var exception = assertThrows(IllegalStateException.class,
                () -> new UserService(userRepository, passwordEncoder, cacheManager));

        assertEquals(
                "Cache 'RECAPTCHA_TOKENS' is not configured. Application cannot start.",
                exception.getMessage());
    }

    // START createUser

    @ParameterizedTest
    @MethodSource("no.vicx.backend.user.UserTestUtils#mockMultipartFileProvider")
    void createUser_givenValidUser_shouldCreateUserInDatabase(MockMultipartFile imageFile) throws IOException {
        var expectedUser = createValidVicxUser();
        expectedUser.setPassword(VALID_BCRYPT_PASSWORD);

        when(passwordEncoder.encode(VALID_PLAINTEXT_PASSWORD)).thenReturn(VALID_BCRYPT_PASSWORD);
        when(userRepository.save(any())).thenReturn(expectedUser);

        sut.createUser(VALID_USER_VM, imageFile);

        verify(passwordEncoder).encode(VALID_PLAINTEXT_PASSWORD);

        var userCaptor = ArgumentCaptor.forClass(VicxUser.class);
        verify(userRepository).save(userCaptor.capture());

        var capturedUser = userCaptor.getValue();
        assertEquals(VALID_BCRYPT_PASSWORD, capturedUser.getPassword());

        if (imageFile == null || imageFile.isEmpty()) {
            assertNull(capturedUser.getUserImage());
        } else {
            assertNotNull(capturedUser.getUserImage());
        }

        verify(reCaptchaCache).evictIfPresent(VALID_USER_VM.recaptchaToken());
    }

    // START getUser

    @Test
    void getUser_givenNonExistingUser_expectException() {
        when(userRepository.findByUsername("user1")).thenReturn(Optional.empty());

        assertThrows(NotFoundException.class, () -> sut.getUserByUserName("user1"));
    }

    @Test
    void getUser_givenExistingUser_expectUser() {
        var userInTest = createValidVicxUser();

        when(userRepository.findByUsername("user1")).thenReturn(Optional.of(userInTest));

        var userInDb = sut.getUserByUserName("user1");

        assertNotNull(userInDb);
    }

    // START updateUser

    @Test
    void updateUser_givenExistingUser_shouldUpdateUserInDatabase() {
        var patchVm = new UserPatchVm("~name~", "foo@bar.com");
        var userInTest = createValidVicxUser();

        when(userRepository.findByUsername("user1")).thenReturn(Optional.of(userInTest));
        when(userRepository.save(any())).thenReturn(userInTest);

        sut.updateUser(patchVm, "user1");

        verify(userRepository).save(userInTest);
    }

    // START isValidPassword

    @Test
    void isValidPassword_givenNonExistingUser_expectException() {
        when(userRepository.findByUsername("user1")).thenReturn(Optional.empty());

        assertThrows(NotFoundException.class, () ->
                sut.isValidPassword("user1", "~clear-text-password~"));
    }

    @ParameterizedTest
    @ValueSource(booleans = {true, false})
    void isValidPassword(boolean expected) {
        when(userRepository.findByUsername("user1")).thenReturn(Optional.of(createValidVicxUser()));
        when(passwordEncoder.matches(VALID_PLAINTEXT_PASSWORD, VALID_BCRYPT_PASSWORD)).thenReturn(expected);

        assertEquals(expected, sut.isValidPassword("user1", VALID_PLAINTEXT_PASSWORD));
    }

    // START updatePassword

    @Test
    void updatePassword_givenExistingUser_() {
        var changePasswordVm = new ChangePasswordVm("~current-password~", VALID_PLAINTEXT_PASSWORD);
        var userInTest = createValidVicxUser();

        when(passwordEncoder.encode(VALID_PLAINTEXT_PASSWORD)).thenReturn(VALID_BCRYPT_PASSWORD);
        when(userRepository.findByUsername("user1")).thenReturn(Optional.of(userInTest));
        when(userRepository.save(any())).thenReturn(userInTest);

        sut.updatePassword(changePasswordVm, "user1");

        verify(userRepository).save(userInTest);
    }
}

File: backend-spring-boot/src/test/kotlin/no/vicx/backend/user/service/UserServiceTest.kt
package no.vicx.backend.user.service

import io.kotest.assertions.assertSoftly
import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.BehaviorSpec
import io.kotest.data.Row1
import io.kotest.data.forAll
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.slot
import io.mockk.verify
import no.vicx.backend.error.NotFoundException
import no.vicx.backend.user.UserTestUtils.VALID_USER_VM
import no.vicx.backend.user.UserTestUtils.createMockMultipartFile
import no.vicx.backend.user.UserTestUtils.createValidVicxUser
import no.vicx.backend.user.vm.ChangePasswordVm
import no.vicx.backend.user.vm.UserPatchVm
import no.vicx.database.user.UserRepository
import no.vicx.database.user.VicxUser
import org.springframework.cache.Cache
import org.springframework.cache.CacheManager
import org.springframework.http.MediaType
import org.springframework.security.crypto.password.PasswordEncoder
import java.util.Optional

class UserServiceTest : BehaviorSpec({
    val userRepository: UserRepository = mockk()
    val passwordEncoder: PasswordEncoder = mockk()
    val cacheManager: CacheManager = mockk()
    val reCaptchaCache: Cache = mockk(relaxed = true)

    lateinit var sut: UserService

    val userSlot = slot<VicxUser>()

    beforeContainer {
        every { cacheManager.getCache("RECAPTCHA_TOKENS") } returns reCaptchaCache
        every { userRepository.findByUsername("user1") } returns Optional.of(createValidVicxUser())
        every { passwordEncoder.encode(VicxUser.VALID_PLAINTEXT_PASSWORD) } returns VicxUser.VALID_BCRYPT_PASSWORD
        every { userRepository.save(capture(userSlot)) } answers {
            userSlot.captured
        }

        sut = UserService(
            userRepository = userRepository,
            passwordEncoder = passwordEncoder,
            cacheManager = cacheManager,
        )
    }

    Given("UserService construction without reCAPTCHA cache") {
        beforeContainer {
            every { cacheManager.getCache("RECAPTCHA_TOKENS") } returns null
        }

        When("constructing UserService") {
            val thrown = shouldThrow<IllegalArgumentException> {
                UserService(
                    userRepository = userRepository,
                    passwordEncoder = passwordEncoder,
                    cacheManager = cacheManager,
                )
            }

            Then("exception should contain expected message") {
                thrown.message shouldBe "Cache 'RECAPTCHA_TOKENS' is not configured. Application cannot start."
            }
        }
    }

    Given("createUser") {
        When("creating a new user without user image") {
            sut.createUser(VALID_USER_VM, null)

            Then("should create a new user") {
                assertSoftly(userSlot.captured) {
                    name shouldBe VALID_USER_VM.name
                    userImage shouldBe null
                }

                verify {
                    passwordEncoder.encode(VicxUser.VALID_PLAINTEXT_PASSWORD)
                    reCaptchaCache.evictIfPresent(any())
                }
            }
        }

        When("creating a new user with user image") {
            sut.createUser(
                createUserVm = VALID_USER_VM,
                image = createMockMultipartFile(
                    "test-png.png",
                    MediaType.IMAGE_PNG_VALUE
                )
            )

            Then("should create a new user") {
                userSlot.captured.userImage shouldNotBe null

                verify {
                    passwordEncoder.encode(VicxUser.VALID_PLAINTEXT_PASSWORD)
                    reCaptchaCache.evictIfPresent(any())
                }
            }
        }
    }

    Given("getUserByUserName") {
        When("getting user by username for existing user") {
            val user = sut.getUserByUserName("user1")

            Then("it should return a User") {
                user shouldNotBe null
            }
        }

        When("getting user by username for non-existing user") {
            every { userRepository.findByUsername("user1") } returns Optional.empty()

            val thrown = shouldThrow<NotFoundException> {
                sut.getUserByUserName("user1")
            }

            Then("expect user not found error") {
                thrown.message shouldBe "User user1 not found"
            }
        }
    }

    Given("updateUser") {
        val patchVm = UserPatchVm("~name~", "foo@bar.com")

        When("updating user") {
            sut.updateUser(patchVm, "user1")

            Then("expect user to be updated") {
                assertSoftly(userSlot.captured) {
                    name shouldBe patchVm.name
                    email shouldBe patchVm.email
                }

                verify { userRepository.save(any()) }
            }
        }
    }

    Given("isValidPassword") {
        forAll(
            Row1(true),
            Row1(false)
        ) { isValidPassword ->
            every { passwordEncoder.matches(any(), any()) } returns isValidPassword

            When("checking if password is valid: $isValidPassword") {
                val result = sut.isValidPassword("user1", "~password~")

                Then("result should be") {
                    result shouldBe isValidPassword

                    verify { passwordEncoder.matches(any(), any()) }
                }
            }
        }
    }

    Given("updatePassword") {
        val changePasswordVm = ChangePasswordVm(
            currentPassword = "~current-password~",
            password = VicxUser.VALID_PLAINTEXT_PASSWORD
        )

        When("updating password") {
            sut.updatePassword(changePasswordVm, "user1")

            Then("expect user to be updated") {
                verify {
                    passwordEncoder.encode(changePasswordVm.password)
                    userRepository.save(any())
                }
            }
        }
    }
})
